File: 0959/solution.py
class Solution:
    def regionsBySlashes(self, grid: list[str]) -> int:
        N = len(grid)
        new_grid = [[0 for _ in range(N * 3)] for _ in range(N * 3)]

        for r in range(N):
            for c in range(N):
                if grid[r][c] == "/":
                    new_grid[r * 3][c * 3 + 2] = 1
                    new_grid[r * 3 + 1][c * 3 + 1] = 1
                    new_grid[r * 3 + 2][c * 3] = 1
                elif grid[r][c] == "\\":
                    new_grid[r * 3][c * 3] = 1
                    new_grid[r * 3 + 1][c * 3 + 1] = 1
                    new_grid[r * 3 + 2][c * 3 + 2] = 1

        def dfs(r, c):
            if (
                r < 0
                or r >= len(new_grid)
                or c < 0
                or c >= len(new_grid)
                or new_grid[r][c] == 1
            ):
                return
            new_grid[r][c] = 1
            dfs(r + 1, c)
            dfs(r - 1, c)
            dfs(r, c + 1)
            dfs(r, c - 1)

        res = 0
        for r in range(len(new_grid)):
            for c in range(len(new_grid)):
                if new_grid[r][c] == 0:
                    res += 1
                    dfs(r, c)

        return res

File: 0959/solution.java
class Solution {
   public int regionsBySlashes(String[] grid) {
      int N = grid.length;
      int[][] newGrid = new int[N * 3][N * 3];

      for (int r = 0; r < N; r++) {
         for (int c = 0; c < N; c++) {
            char ch = grid[r].charAt(c);
            if (ch == '/') {
               newGrid[r * 3][c * 3 + 2] = 1;
               newGrid[r * 3 + 1][c * 3 + 1] = 1;
               newGrid[r * 3 + 2][c * 3] = 1;
            } else if (ch == '\\') {
               newGrid[r * 3][c * 3] = 1;
               newGrid[r * 3 + 1][c * 3 + 1] = 1;
               newGrid[r * 3 + 2][c * 3 + 2] = 1;
            }
         }
      }

      int res = 0;
      for (int r = 0; r < N * 3; r++) {
         for (int c = 0; c < N * 3; c++) {
            if (newGrid[r][c] == 0) {
               res++;
               dfs(newGrid, r, c, N * 3);
            }
         }
      }

      return res;
   }

   private void dfs(int[][] grid, int r, int c, int N) {
      if (r < 0 || r >= N || c < 0 || c >= N || grid[r][c] == 1) {
         return;
      }
      grid[r][c] = 1;
      dfs(grid, r + 1, c, N);
      dfs(grid, r - 1, c, N);
      dfs(grid, r, c + 1, N);
      dfs(grid, r, c - 1, N);
   }
}

// time O(n^2)
// space O(n^2)